Macros 2.0 (v0.4) - Add scoped macros (last arg can be scoped), `{{if}}` macro and macro flags (baseline implementation)

Wolfsblvt opened the pull request:
Continuation of #4820, enhancing the new macro engine with new features I had already planned, that didn't make it in the first throw of the base PR.

This time, there are actually visible and usable new features - besides the nested macros from before, and the fixed execution order.
Mostly, this introduces:

- **Scoped Macros** - Allow macros to build a scope with a closing text, to make defining multiline arguments easier.
- **`{{if}}` Macro** - Introduced the long-awaited `{{if}}` macro to allow conditionals inside macro definitions.
- **Autocomplete Improvments** - Slashcommand autocomplete for macros got better.

More features will follow soon...

-----

### New Features

**Scoped Macro Syntax**
- New opening/closing tag syntax: `{{macro}}content{{/macro}}`
- Scoped content is passed as the last unnamed argument, and will be trimmed (including indentation)
- Works with any macro that can accept an additional argument

**Macro Execution Flags**
- Flag prefix syntax: `{{!macro}}`, `{{?macro}}`, etc.
- `/` - Used for closing a scoped macro
- `#` - "Preserve Whitespaces" flag, to preserve whitespaces in scoped blocks. Also legacy usage of Handlebar-style if blocks.
- Most are still unimplemented, but already added as baseline

**Conditional Macro: `{{if}}`**
- Conditional content rendering: `{{if condition}}content{{/if}}`
- Auto-resolves macro names as conditions (e.g., `{{if personality}}`)
- Supports `{{else}}` branches: `{{if x}}then{{else}}otherwise{{/if}}`
- Negation with `!` prefix: `{{if !personality}}`
- Will trim the respective branches, so it allows clean multiline if/else definitions
- Falsy values: empty string, `"false"`, `"off"`, `"0"`

**Scoped `{{trim}}` Macro**
- New scoped usage: `{{trim}}  content  {{/trim}}` → `content`
- Non-scoped behavior unchanged (post-processing marker)

### Autocomplete Improvements

- **Flag suggestions** with icons and descriptions
- **Closing tag suggestions** when inside unclosed scoped macros
- **Arity warnings** when argument count is invalid
- **Scoped content hints** showing which macro's content you're editing
- **`{{if}}` condition autocomplete** - suggests zero-arg macros without braces
- **`{{else}}` awareness** - only selectable inside `{{if}}` blocks
- **Whitespace preservation** - matches opening padding when auto-closing

### Other Improvements

- Centralized macro identifier validation with pattern enforcement
- Does not allow registering macros with invalid identifiers
- Fixed nested macro indexing for proper autocomplete inside nested macros
- `rawOriginal` and `rawArgs` available in macro execution context
- Allow extensions to add custom pre- and post-processors to the `MacroEngine`

Combined, the new features allow stuff like this:
```
{{ if {{getvar::custom-block}} }}
  # Custom Block Content
  The following will be used as custom content:

  {{getvar::custom-block}}
  {{setvar::custom-block-added::true}}
{{ /if }}
```
```
{{ setvar multine-text }}
This is the first line.
Second line.
{{ /setvar }}
```

## Copilot Summary
This pull request introduces several improvements and new features to the macro engine, macro autocomplete, and core macro definitions. The most significant updates include the addition of macro execution flags (with supporting infrastructure and documentation), enhancements to the macro lexer for better identifier and flag handling, new utility macros (`if`/`else` and improved `trim`), and improvements to the autocomplete sorting and UI. These changes lay the groundwork for more flexible and powerful macro usage, while also improving the user experience in macro editing.

**Macro Engine Enhancements:**

* Added a new `MacroFlags` system to support macro execution flags (such as `!`, `?`, `~`, `>`, `/`, `#`, `.`, `), including definitions, parsing, validation, and utility functions in `MacroFlags.js`. This enables future features like immediate/delayed evaluation, filters, and variable shorthands.
* Updated the macro lexer (`MacroLexer.js`) to recognize and document all macro execution flags, split out the filter flag for special handling, and standardized macro identifier patterns for both lexer and validation. [[1]](diffhunk://#diff-2827513de6ad694faaf2815fd7ac188abe20d29f8986e733576da8a9e0aca7c1R6-R18) [[2]](diffhunk://#diff-2827513de6ad694faaf2815fd7ac188abe20d29f8986e733576da8a9e0aca7c1L28-R58) [[3]](diffhunk://#diff-2827513de6ad694faaf2815fd7ac188abe20d29f8986e733576da8a9e0aca7c1L48-R82)

**Core Macro Feature Additions:**

* Introduced a new scoped `trim` macro and a powerful `if`/`else` conditional macro, enabling conditional content and trimming inside macro blocks. The `if` macro supports macro name resolution, negation, and an internal marker for `else` branches.
* Integrated the new macro flags infrastructure into macro registration and handler logic.

**Autocomplete and UI Improvements:**

* Added `sortPriority` to `AutoCompleteOption` and updated sorting logic in `AutoComplete.js` to prioritize options by this value before fuzzy/alphabetical sorting, ensuring more relevant suggestions appear first. [[1]](diffhunk://#diff-8d64231974efb29eb1d9bacd4c9785e67f51e27ef9c57c56b7a4c0579cb2f0aaR16-R21) [[2]](diffhunk://#diff-cf97a1d6c169a2a266ab0a190d0214b400cf1ea5ae43773d0e4c1c6a3028c274L396-R409)
* Improved autocomplete UI and details panels for macros, including new CSS for warning/info banners and closing tag details.
* Enhanced autocomplete trigger logic to also activate when an option is selected, not just on input or forced invocation.

**Other Improvements:**

* Fixed the way `makeNoMatchText` and `makeNoOptionsText` are assigned in `AutoCompleteNameResultBase` for better extensibility.

These changes collectively make the macro system more robust, extensible, and user-friendly, setting the stage for future macro features and improved editing workflows.


<!-- Put X in the box below to confirm -->

## Checklist:

- [x] I have read the [Contribution guidelines](https://www.youtube.com/watch?v=e-GwojpaoQI).

Wolfsblvt committed:
Macros: make category optional, default to UNCATEGORIZED

Wolfsblvt committed:
Tests: Update macro tests to match new error handling behavior

- Change MacroRegistry tests from expecting thrown errors to capturing console.error logs
- Update MacroLexer tests to expect plaintext fallback instead of lexer errors for invalid tokens
- Fix MacroEngine test to use `char` macro instead of `newline` for arity validation
- Update MacroParser test with corrected expected error message for invalid identifiers
- Remove "[Error]" prefixes from test descriptions where lexer no longer errors

Cohee1207 committed:
Add 'setspriteoverride' optional 'name' argument

Wolfsblvt committed:
Macros: Implement macro execution flags system

- Add MacroFlags module with flag parsing and validation (!, ?, ~, /, >, ., $, #)
- Update lexer to tokenize flags as separate tokens before macro identifier
- Modify parser to capture flags in CST under 'flags' label
- Update CST walker to parse flag tokens into MacroFlags object
- Pass parsed flags to macro handlers via MacroCall and MacroContext
- Update autocomplete parser to handle flags before identifier
- Add comprehensive tests for flag parsing,

Wolfsblvt committed:
Macros: Fix autocomplete positioning for macros with flags and whitespace

- Add identifierStart to parseMacroContext to track where identifier begins in macro text
- Update autocomplete to use identifierStart for correct range calculation
- Simplify indexMacros regex to use global flag instead of manual loop
- Use parseMacroContext in indexMacros to extract identifier (handles flags/whitespace)
- Fix autocomplete range starting at wrong position when flags or whitespace present

Wolfsblvt committed:
Macros: Add autocomplete support for macro execution flags

- Add MacroFlagAutoCompleteOption class for rendering flag options in autocomplete
- Extend parseMacroContext to track currentFlag (flag cursor is on) and isInFlagsArea
- Track flagEndPositions to determine which flag cursor is currently typing
- Update #buildEnhancedMacroOptions to show flag options when cursor is in flags area
- Show current flag first if cursor just typed it, then show remaining available flags
- Add renderItem and renderDetails methods to MacroFlag

Cohee1207 committed:
Refactor ElevenLabs TTS API key handling (#4906)

* Refactor ElevenLabs TTS API key handling #4483

* Remove unused connection button and related event handler from ElevenLabs TTS provider

* Add ElevenLabs STT endpoint

chungchandev committed:
Add caching system prompt feature for OpenRouter Gemini (#4903)

* feat: add caching system prompt for OpenRouter Gemini

* fix: resolve reviews

Wolfsblvt committed:
Macros: Implement scoped macro syntax with opening and closing tags

- Add scoped macro processing to MacroCstWalker to find and merge opening/closing pairs
- Parse closing block flag (/) to identify closing macros and match with opening tags
- Extract content between opening and closing tags as the last unnamed argument
- Add `isScoped` property to MacroCall and MacroContext to track scoped invocations
- Implement `#processScopedMacros` to find outermost matching pairs and handle nesting

Wolfsblvt committed:
Macros: Add autocomplete warnings, scoped content info, and closing tag suggestions

- Add arity warning banners in autocomplete details for invalid argument counts
- Show warning when using space-separated args on multi-arg or no-arg macros
- Add scoped content info banner when cursor is inside unclosed scoped macro
- Implement MacroClosingTagAutoCompleteOption to suggest closing tags for scoped macros
- Add sortPriority property to AutoCompleteOption for controlling sort order

Wolfsblvt committed:
SlashCommands: Disable unimplemented flags and closing flag when no unclosed scopes in autocomplete

- Set closing block flag as non-selectable when no unclosed scopes exist
- Set unimplemented flags as non-selectable with empty valueProvider
- Lower sort priority (12) for non-selectable flags vs selectable flags (10)

Cohee1207 committed:
Update GitHub links to llama.cpp

Wolfsblvt committed:
Autocomplete: Show scoped content info and auto-close no-arg macros

- Show scoped content info when cursor is at closing }} of unclosed scoped macro
- Auto-complete no-arg macros with closing }} using valueProvider
- Trigger autocomplete on select (isSelect) to refresh after choosing an option
- Simplify MacroFlagAutoCompleteOption to use base makeItem for consistent styling
- Change closing tag icon from '{/}' to '{/' for better visual consistency

Wolfsblvt committed:
Macros: Add scoped trim macro to trim content inside opening/closing tags

- Add scoped usage for {{trim}}content{{/trim}} to trim whitespace from content
- Keep non-scoped {{trim}} behavior (post-processing marker) for backward compatibility
- Add optional unnamed 'content' argument for scoped usage
- Update description to explain both scoped and non-scoped behavior
- Handler checks isScoped flag to determine which behavior to use

Wolfsblvt committed:
Autocomplete: Fix closing tag parsing to prevent `/` being treated as flag

- Add special case in parseMacroContext to detect closing tags (`/` + identifier char)
- Stop flag parsing when `/` is followed by identifier character (closing tag syntax)
- Simplify MacroClosingTagAutoCompleteOption valueProvider to return full closing tag
- Remove input-based logic since autocomplete replaces entire identifier

my-alt-acct committed:
Add model selection support for llama.cpp router mode (#4910)

* Add model selection support for llama.cpp router mode

- Add llamacpp_model setting to textgen-settings.js
- Implement loadLlamaCppModels() function to fetch and populate models
- Add onLlamaCppModelSelect() handler for model selection
- Update status check to load llama.cpp models when connecting
- Update getTextGenModel() to return selected llama.cpp model
- Add model dropdown to HTML UI in llama.cpp section
- Initialize event handlers and Select2 for better UX
- Add llamacpp_model to preset manager for save/load support
- Add llamacpp_model to slash commands support

This implements model selection for llama.cpp router mode, allowing
users to select from multiple models without restarting the server.
Follows the same pattern as Ollama, Tabby, and vLLM implementations.

* Correct spelling

* Fix clear selection position

---------

Co-authored-by: Cohee <18619528+Cohee1207@users.noreply.github.com>

Wolfsblvt committed:
Macros: Add {{if}} conditional macro with auto-resolution of macro names

- Add {{if condition}}content{{/if}} macro to conditionally show content
- Auto-resolve condition if it matches a registered macro name (0 required args)
- Support both scoped content ({{if x}}...{{/if}}) and explicit args ({{if::x::content}})
- Treat empty string, "false", "off", "0" as falsy conditions
- Inherit environment context when resolving macro names
- Update autocomplete warning to allow space-separated syntax

Wolfsblvt committed:
Macros: Add centralized identifier validation with pattern enforcement

- Export MACRO_IDENTIFIER_PATTERN from MacroLexer for reuse across modules
- Add isIdentifierValid() helper function to validate macro names and aliases
- Enforce identifier pattern: must start with letter, followed by word chars or hyphens
- Update macro registration to validate both primary names and alias identifiers
- Improve error messages to explain identifier requirements
- Add comprehensive e2e tests for valid/invalid identifier patterns

Wolfsblvt committed:
Macros: Add tests for scoped {{trim}} macro functionality

Wolfsblvt committed:
SlashCommands: Fix macro indexing to properly handle nested macros with brace depth tracking

- Replace regex-based macro detection with manual brace depth tracking
- Track opening/closing brace pairs to correctly identify macro boundaries
- Ensure nested macros like {{reverse::Hey {{user}}}} are properly indexed
- Index both outer and inner macros by scanning content recursively
- Handle unclosed macros by defaulting to end of text

Wolfsblvt committed:
Macros: Add {{else}} branch support to {{if}} conditional macro

- Add {{else}} macro as marker to split then/else branches in {{if}} blocks
- Use control character sequence (\u0000\u001FELSE\u001F\u0000) as internal marker
- Split scoped content on else marker and trim both branches independently
- Return then-branch if condition is truthy, else-branch if falsy
- Auto-suggest {{else}} in autocomplete when inside scoped {{if}} block
- Make {{else}} non-selectable in autocomplete when outside {{if}} scope

Wolfsblvt committed:
Macros: Add negation support to {{if}} conditional macro with ! prefix

- Add ! prefix support to invert condition evaluation in {{if}} macro
- Parse original macro text to detect ! prefix before macro resolution
- Strip ! from condition after detecting inversion to avoid double-negation
- Invert isFalsy result when ! prefix is detected in original condition
- Prevent ! in resolved values from triggering inversion (only original syntax)

Cohee1207 committed:
Add glm-4.7 model option and context mapping for Z.AI

Wolfsblvt committed:
Autocomplete: Add {{if}} condition autocomplete with zero-arg macro suggestions

- Add EnhancedMacroAutoCompleteOptions typedef for noBraces/paddingAfter/closeWithBraces options
- Support options object in EnhancedMacroAutoCompleteOption constructor alongside context
- Add noBraces mode to display macro names without {{ }} braces (for use as values)
- Add paddingAfter option to match opening whitespace style before closing }}

Wolfsblvt committed:
Autocomplete: Match opening whitespace padding when auto-closing macros

Wolfsblvt committed:
Fix `{{if}}` example usages

Wolfsblvt committed:
Macros: Hide `comment` alias from autocomplete suggestions for `//` macro

Cohee1207 committed:
Add select2 style for NanoGPT list
Closes #4911

Wolfsblvt committed:
Macros: Simplify {{trim}} handler with destructured parameters and clearer content check

Wolfsblvt commented:
@codex review

Cohee1207 commented:
@codex review

Wolfsblvt committed:
Macros: Use MacroEngine.evaluate for zero-arg macro resolution in {{if}} condition handler

Cohee1207 commented on public/scripts/macros/definitions/core-macros.js:
```diff
@@ -59,11 +60,114 @@ export function registerCoreMacros() {
+        handler: (env) => {
+            if (env.isScoped && env.unnamedArgs.length > 0) {
+                // Scoped usage: trim the content inside
+                return env.unnamedArgs[0].trim();
```
`{{trim}}` would have different meaning depending on its position. Unscoped `{{trim}}` only trims newlines, while scopes will also trim whitespace.

Wolfsblvt replied:
Yes, that was the intention. Just needs to be documented.

Wolfsblvt replied:
Let me know if we need more documentation there than the one currently already existing on the `{{trim}}` macro itself.

Cohee1207 commented on public/scripts/macros/definitions/core-macros.js:
```diff
@@ -59,11 +60,114 @@ export function registerCoreMacros() {
+            },
+            {
+                name: 'content',
+                description: 'The content to return if condition is truthy (typically provided as scoped content). May contain {{else}} to define an else branch. Both branches are trimmed.',
```
> Both branches are trimmed

This is a problem for Handlebars compatibility. It preserves newlines nad whitespace, hence a weird
```
{{#if foo}}{{foo}}
{{/if}}
```

Wolfsblvt replied:
As discussed, this will be handled as a macro flag (we just use this `#`) that allows scoped macros to preserve whitespaces.
This both keeps Handlebars compatibility, and also allows any other scoped macro to preserve those, without weird `{{noop}}` workarounds.

Wolfsblvt replied:
`#` is no-trim for scoped blocks now. So solved.

Cohee1207 commented on public/scripts/macros/definitions/core-macros.js:
```diff
@@ -59,11 +60,114 @@ export function registerCoreMacros() {
+    // Condition can be a macro name (resolved automatically) or any value
+    MacroRegistry.registerMacro('if', {
+        category: MacroCategory.UTILITY,
+        description: 'Conditional macro. Returns the trimmed content if the condition is truthy, otherwise returns nothing (or the trimmed else branch if present). Prefix the condition with ! to invert. If the condition is a registered macro name (without braces), it will be resolved first.',
```
> Returns the trimmed content

I think it's a problem. See below. I wouldn't expect it to do implicit whitespace manipulation.

Wolfsblvt replied:
Special documentation or handling of trimming in `{{if}}` macro removed. It just does the same as all other scoped macros now.
(Only small quirk is that we trim around the `{{else}}` block if the whole scope should be trimmed.
if macro is an exception, but having something like two scopes inside.

Cohee1207 commented on public/scripts/macros/definitions/core-macros.js:
```diff
@@ -59,11 +60,114 @@ export function registerCoreMacros() {
+    // Only meaningful inside a scoped {{if}} macro
+    MacroRegistry.registerMacro('else', {
+        category: MacroCategory.UTILITY,
+        description: 'Marks the else branch inside a scoped {{if}} block. Only works inside {{if}}...{{/if}}. If used outside, returns an invisible marker.',
```
Is there any expected use case for the invisible marker? Maybe return empty string if unscoped?

Wolfsblvt replied:
Nested macros resolve first. Without a marker, there is no way to split the scoped content into both branches. If it replaces empty, it's just gone.
We *could* have this macro be special by running the actual content block run through macro evaluation **again**, but that just... looks weird in my opinion.
Using a marker is cleaner and easier. And if you use `{{else}}` outside of a scoped if... well, then that's your fault.

Wolfsblvt replied:
If you got a better idea, let me know. Like I said, we could re-evaluate the correct block after splitting the raw content, but I'd say that's less clear and could lead to other issues.

Cohee1207 replied:
Idea: remove it in post-processing. Cause it has no business being in the output string.

Wolfsblvt replied:
Is another post processing regex that has to run. But shouldn't be too bad. Can do that.

Wolfsblvt replied:
Done.

Cohee1207 commented on public/scripts/macros/engine/MacroParser.js:
```diff
@@ -23,7 +23,7 @@ class MacroParser extends CstParser {
     constructor() {
         super(MacroLexer.def, {
-            traceInitPerf: true,
+            //traceInitPerf: true,
```
Should we backport that to staging before releasing?

Wolfsblvt replied:
Yes, we should. It clutters the console on page load. Forgot to remove it before finishing the other PR. It's really not needed during production use. For both Lexer and Parser.

Cohee1207 replied:
Ok, I did that. If you need to bring it back locally, maybe add a debug menu option.

https://github.com/SillyTavern/SillyTavern/commit/432639e341941686ab2afed869abcd0368a69974

Wolfsblvt replied:
Yeah, makes sense to just have a toggle. Debug toggle going into localStorage, or something. Or... I just remember to turn it off again whenever I work on performance stuff.

Cohee1207 commented on public/scripts/macros/engine/MacroRegistry.js:
```diff
@@ -179,7 +188,7 @@ class MacroRegistry {
 
         try {
-            if (typeof name !== 'string' || !name) throw new Error('Macro name must be a non-empty string');
+            if (!isIdentifierValid(name)) throw new Error(`Macro name "${name}" is invalid. Must start with a letter, followed by word chars or hyphens.`);
```
Not sure about the "word chars" terminology, it requires knowledge of Unicode character classes.

Wolfsblvt replied:
A better suggestion how to word it? "alphanumeric"?

Cohee1207 replied:
That was my thought.

Wolfsblvt replied:
Done.

Cohee1207 committed:
Disable macro engine init traces

Wolfsblvt commented on public/scripts/macros/definitions/core-macros.js:
```diff
@@ -59,11 +60,114 @@ export function registerCoreMacros() {
+                    name: condition,
+                    args: [],
+                    unnamedArgs: [],
+                    list: null,
```
Fixed by actually using the engine evaluation instead of calling handler or executeMacro.

Wolfsblvt committed:
Macros: Add auto-trim for scoped content with # flag to preserve whitespace

- Auto-trim scoped content by default in MacroCstWalker before passing to handlers
- Add preserveWhitespace flag (# symbol) to prevent auto-trimming when needed
- Rename legacyHash flag to preserveWhitespace across engine and definitions
- Update {{trim}} handler to rely on engine auto-trim for scoped content
- Update {{if}} handler to respect # flag when trimming branches around {{else}} marker

Wolfsblvt committed:
Macros: Clarify macro name validation error message to use "alphanumeric characters" instead of "word chars"

Wolfsblvt committed:
Merge branch 'staging' into macros-2.0-v0.4

Cohee1207 committed:
Update type annotations for instruct presets and context presets

Cohee1207 committed:
Fix test for new error text

Cohee1207 committed:
Add MacroStoryString tests

Wolfsblvt committed:
Merge branch 'staging' into macros-2.0-v0.4

Wolfsblvt committed:
Add trimContent utility for consistent indentation dedenting in scoped macros

- Add trimScopedContent method to MacroEngine that trims and dedents scoped content based on first non-empty line indentation
- Pass trimContent utility through evaluation context to all macro handlers
- Update {{if}} macro to use trimContent instead of direct trim() call
- Update auto-trim logic in MacroCstWalker to use trimContent for consistent dedenting
- Add trimContent to MacroExecutionContext type definitions

Wolfsblvt committed:
Add ELSE_MARKER export and cleanup leftover markers in macro processing

Wolfsblvt committed:
Update trimContent parameter to use options object pattern in JSDoc

Wolfsblvt committed:
Add processor registration system to MacroEngine with priority-based execution

- Add MacroProcessor callback and RegisteredProcessor typedef for pre/post processors
- Add addPreProcessor/removePreProcessor and addPostProcessor/removePostProcessor methods with priority-based sorting
- Refactor core legacy syntax handling into registered processors with reserved priorities (0-50)
- Move legacy time syntax, marker replacements, brace unescaping, trim macro, and ELSE_MARKER cleanup to registered processors

Wolfsblvt committed:
Split core processor registration into separate pre and post processor methods
